File: apps/ios/Barns/BarnsApp/Presentation/Support/ConsultationDraftViewModel.swift
import Combine
import Foundation

@MainActor
final class ConsultationDraftViewModel: ObservableObject {
    @Published var topic: String = ""
    @Published var body: String = ""
    @Published var category: ConsultationCategory = .maintenance
    @Published var urgency: ConsultationUrgency = .normal
    @Published private(set) var isSaving = false
    @Published private(set) var savedAt: Date?

    // Selectable options for the local-only draft form.
    let categories: [ConsultationCategory] = [.maintenance, .care, .replacement, .other]
    let urgencies: [ConsultationUrgency] = [.low, .normal, .high]

    /// When the draft is started from a registered greenery item detail, this
    /// holds the item context so the note can be prefilled. Optional so the
    /// general Support draft flow keeps working unchanged.
    private let itemContext: ProductItem?
    /// Human-readable name of the registered greenery this note is about.
    let itemContextName: String?

    private var existing: ConsultationDraft?
    private let getConsultationDraftUseCase: GetConsultationDraftUseCase
    private let saveConsultationDraftUseCase: SaveConsultationDraftUseCase

    init(
        getConsultationDraftUseCase: GetConsultationDraftUseCase,
        saveConsultationDraftUseCase: SaveConsultationDraftUseCase,
        item: ProductItem? = nil
    ) {
        self.getConsultationDraftUseCase = getConsultationDraftUseCase
        self.saveConsultationDraftUseCase = saveConsultationDraftUseCase
        self.itemContext = item
        self.itemContextName = item?.name
    }

    var canSave: Bool {
        !topic.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && !isSaving
    }

    func load() async {
        // Item-specific note: prefill from the registered greenery context.
        // Local-only; nothing is submitted anywhere.
        if let item = itemContext {
            if topic.isEmpty {
                topic = "Consultation: \(item.name)"
            }
            if body.isEmpty {
                body = Self.contextSummary(for: item)
            }
            return
        }

        guard let draft = try? await getConsultationDraftUseCase.execute() else { return }
        existing = draft
        topic = draft.topic
        body = draft.body
        category = draft.category
        urgency = draft.urgency
    }

    /// Builds a local-only context summary from existing item fields so the
    /// user can add their concern below it before contacting support.
    private static func contextSummary(for item: ProductItem) -> String {
        let display = ProductItemPresentation(item: item)
        return """
        Item: \(display.name)
        Type: \(display.typeLabel)
        Category: \(display.categoryLabel)
        Location: \(display.locationLabel)
        Status: \(display.statusLabel)

        Concern:\u{0020}
        """
    }

    /// Saves the draft locally only. It is never sent to a server.
    func save() async {
        guard canSave else { return }
        isSaving = true
        defer { isSaving = false }
        let saved = try? await saveConsultationDraftUseCase.execute(
            existing: existing,
            productItemId: itemContext?.id,
            topic: topic.trimmingCharacters(in: .whitespacesAndNewlines),
            category: category,
            urgency: urgency,
            body: body.trimmingCharacters(in: .whitespacesAndNewlines)
        )
        if let saved {
            existing = saved
            savedAt = saved.updatedAt
        }
    }
}

File: apps/android/app/src/main/java/com/barns/app/presentation/support/ConsultationDraftViewModel.kt
package com.barns.app.presentation.support

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.barns.app.domain.model.ConsultationCategory
import com.barns.app.domain.model.ConsultationDraft
import com.barns.app.domain.model.ConsultationUrgency
import com.barns.app.domain.model.ProductItem
import com.barns.app.domain.usecase.support.GetConsultationDraftUseCase
import com.barns.app.domain.usecase.support.SaveConsultationDraftUseCase
import com.barns.app.presentation.myitems.ProductItemPresentation
import java.time.Instant
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class ConsultationDraftViewModel(
    private val getConsultationDraftUseCase: GetConsultationDraftUseCase,
    private val saveConsultationDraftUseCase: SaveConsultationDraftUseCase,
    // When the draft is started from a registered greenery item detail, this
    // holds the item context so the note can be prefilled. Optional so the
    // general Support draft flow keeps working unchanged.
    private val item: ProductItem? = null,
) : ViewModel() {
    /** Human-readable name of the registered greenery this note is about. */
    val itemContextName: String? = item?.name

    // Selectable options for the local-only draft form.
    val categories: List<ConsultationCategory> = listOf(
        ConsultationCategory.MAINTENANCE,
        ConsultationCategory.CARE,
        ConsultationCategory.REPLACEMENT,
        ConsultationCategory.OTHER,
    )
    val urgencies: List<ConsultationUrgency> = listOf(
        ConsultationUrgency.LOW,
        ConsultationUrgency.NORMAL,
        ConsultationUrgency.HIGH,
    )

    private var existing: ConsultationDraft? = null

    private val _state = MutableStateFlow(ConsultationDraftUiState())
    val state: StateFlow<ConsultationDraftUiState> = _state.asStateFlow()

    fun onTopicChange(value: String) {
        _state.value = _state.value.copy(topic = value)
    }

    fun onBodyChange(value: String) {
        _state.value = _state.value.copy(body = value)
    }

    fun onCategoryChange(value: ConsultationCategory) {
        _state.value = _state.value.copy(category = value)
    }

    fun onUrgencyChange(value: ConsultationUrgency) {
        _state.value = _state.value.copy(urgency = value)
    }

    fun load() {
        // Item-specific note: prefill from the registered greenery context.
        // Local-only; nothing is submitted anywhere.
        item?.let { current ->
            val state = _state.value
            _state.value = state.copy(
                topic = if (state.topic.isBlank()) "Consultation: ${current.name}" else state.topic,
                body = if (state.body.isBlank()) contextSummary(current) else state.body,
            )
            return
        }

        viewModelScope.launch {
            val draft = runCatching { getConsultationDraftUseCase.execute() }.getOrNull() ?: return@launch
            existing = draft
            _state.value = _state.value.copy(
                topic = draft.topic,
                body = draft.body,
                category = draft.category,
                urgency = draft.urgency,
            )
        }
    }

    /**
     * Builds a local-only context summary from existing item fields so the
     * user can add their concern below it before contacting support.
     */
    private fun contextSummary(item: ProductItem): String {
        val display = ProductItemPresentation.from(item)
        return buildString {
            appendLine("Item: ${display.name}")
            appendLine("Type: ${display.typeLabel}")
            appendLine("Category: ${display.categoryLabel}")
            appendLine("Location: ${display.locationLabel}")
            appendLine("Status: ${display.statusLabel}")
            appendLine()
            append("Concern: ")
        }
    }

    /** Saves the draft locally only. It is never sent to a server. */
    fun save() {
        val current = _state.value
        if (!current.canSave) return
        _state.value = current.copy(isSaving = true)
        viewModelScope.launch {
            runCatching {
                saveConsultationDraftUseCase.execute(
                    existing = existing,
                    productItemId = item?.id,
                    topic = current.topic.trim(),
                    category = current.category,
                    urgency = current.urgency,
                    body = current.body.trim(),
                )
            }
                .onSuccess { saved ->
                    existing = saved
                    _state.value = _state.value.copy(isSaving = false, savedAt = saved.updatedAt)
                }
                .onFailure {
                    _state.value = _state.value.copy(
                        isSaving = false,
                        errorMessage = "Unable to save the draft. Please try again.",
                    )
                }
        }
    }
}

data class ConsultationDraftUiState(
    val topic: String = "",
    val body: String = "",
    val category: ConsultationCategory = ConsultationCategory.MAINTENANCE,
    val urgency: ConsultationUrgency = ConsultationUrgency.NORMAL,
    val isSaving: Boolean = false,
    val savedAt: Instant? = null,
    val errorMessage: String? = null,
) {
    val canSave: Boolean
        get() = topic.isNotBlank() && !isSaving
}
